sdk/python/feast/constants.py

```diff
@@ -14,142 +14,196 @@
-    # Default Feast Serving URL
-    CONFIG_SERVING_URL_KEY: "localhost:6565",
-    # Enable or disable TLS/SSL to Feast Serving
-    CONFIG_SERVING_ENABLE_SSL_KEY: "False",
-    # Path to certificate(s) to secure connection to Feast Serving
-    CONFIG_SERVING_SERVER_SSL_CERT_KEY: "",
-    # Default connection timeout to Feast Serving, Feast Core, and Feast Job Service (in seconds)
-    CONFIG_GRPC_CONNECTION_TIMEOUT_DEFAULT_KEY: "10",
-    # Default gRPC connection timeout when sending an ApplyFeatureSet command to
-    # Feast Core (in seconds)
-    CONFIG_GRPC_CONNECTION_TIMEOUT_APPLY_KEY: "600",
-    # Time to wait for batch feature requests before timing out.
-    CONFIG_BATCH_FEATURE_REQUEST_WAIT_TIME_SECONDS_KEY: "600",
-    CONFIG_TIMEOUT_KEY: "21600",
-    CONFIG_MAX_WAIT_INTERVAL_KEY: "60",
-    # Authentication Provider - Google OpenID/OAuth
-    CONFIG_AUTH_PROVIDER: "google",
-    CONFIG_SPARK_LAUNCHER: "dataproc",
-    CONFIG_SPARK_INGESTION_JOB_JAR: "https://storage.googleapis.com/feast-jobs/spark/"
-    "ingestion/feast-ingestion-spark-develop.jar",
-    CONFIG_SPARK_STANDALONE_MASTER: "local[*]",
-    CONFIG_REDIS_HOST: "localhost",
-    CONFIG_REDIS_PORT: "6379",
-    CONFIG_REDIS_SSL: "False",
-    CONFIG_SPARK_HISTORICAL_FEATURE_OUTPUT_FORMAT: "parquet",
-    # Enable or disable TLS/SSL to Feast Service
-    CONFIG_JOB_SERVICE_ENABLE_SSL_KEY: "False",
-    # Path to certificate(s) to secure connection to Feast Job Service
-    CONFIG_JOB_SERVICE_SERVER_SSL_CERT_KEY: "",
-    # Disable control loop by default for now
-    CONFIG_JOB_SERVICE_ENABLE_CONTROL_LOOP: "False",
-    CONFIG_STATSD_ENABLED: "False",
-    # IngestionJob DeadLetter Destination
-    CONFIG_DEADLETTER_PATH: "",
-    # ProtoRegistry Address (currently only Stencil Server is supported as registry)
-    # https://github.com/gojekfarm/stencil
-    CONFIG_STENCIL_URL: "",
-}
+class Option:
+    def __init__(self, name, default):
+        self._name = name
+        self._default = default
+
+    def __get__(self, instance, owner):
+        if instance is None:
+            return self._name.lower()
+
+        return self._default
+
+
+class ConfigMeta(type):
+    def __new__(cls, name, bases, attrs):
+        keys = [
+            k
+            for k, v in attrs.items()
+            if not k.startswith("_") and isinstance(v, (str, int, float))
+        ]
+        attrs["__config_keys__"] = keys
+        attrs.update({k: Option(k, attrs[k]) for k in keys})
+        return super().__new__(cls, name, bases, attrs)
+
+
+class ConfigOptions(metaclass=ConfigMeta):
+    """ Feast Configuration Options """
+
+    #: Default datetime column name for point-in-time join
+    DATETIME_COLUMN: str = "datetime"
+
+    #: Environmental variable to specify Feast configuration file location
+    FEAST_CONFIG_FILE_ENV: str = "FEAST_CONFIG"
+
+    #: Default prefix to Feast environmental variables
+    CONFIG_FEAST_ENV_VAR_PREFIX: str = "FEAST_"
```

Comment:
Some of these options should not be defaults. They are more like traditional constants. For example, you cant set this `CONFIG_FEAST_ENV_VAR_PREFIX` because you need it to find the `CONFIG_FEAST_ENV_VAR_PREFIX` config in the first place.

Reply:
Updated.